src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java

```diff
@@ -0,0 +1,63 @@
+package com.wanderersoftherift.wotr.core;
+
+import com.wanderersoftherift.wotr.WanderersOfTheRift;
+import com.wanderersoftherift.wotr.init.WotrAttributes;
+import com.wanderersoftherift.wotr.init.WotrDamageTypes;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.Level;
+import net.neoforged.bus.api.SubscribeEvent;
+import net.neoforged.fml.common.EventBusSubscriber;
+import net.neoforged.neoforge.event.entity.living.LivingIncomingDamageEvent;
+
+@EventBusSubscriber(modid = WanderersOfTheRift.MODID, bus = EventBusSubscriber.Bus.GAME)
+public class OnHitEvent {
+
+    @SubscribeEvent
+    public static void onCriticalHit(LivingIncomingDamageEvent event) {
+        if (event.getSource().getDirectEntity() instanceof LivingEntity livingAttacker) {
+            float dmg = OnHitEffect.critical(event.getAmount(), livingAttacker, event.getEntity(),
+                    event.getEntity().getRandom());
+            event.setAmount(dmg);
+        }
+    }
+
+    @SubscribeEvent
+    public static void onThornsProc(LivingIncomingDamageEvent event) {
+        Level level = event.getEntity().level();
+        DamageSource thorns = new DamageSource(
+                level.registryAccess().lookupOrThrow(Registries.DAMAGE_TYPE).getOrThrow(WotrDamageTypes.THORNS_DAMAGE));
+
+        Entity causer = event.getSource().getEntity();
+        LivingEntity reciever = event.getEntity();
+        if (!WotrDamageTypes.THORNS_DAMAGE.equals(event.getSource().typeHolder()) && causer != null) {
```

Comment:
You should also check if `event.getAmount() == 0`.

Reply:
The equals check always returns false (left side is a ResourceKey, right side is a Holder).

Reply:
Right, it should be:
```suggestion
        if (!event.getSource().typeHolder().is(WotrDamageTypes.THORNS_DAMAGE) && causer != null) {
```

Reply:
Also thorns loop might still be possible if one entity has WotR thorns and the other has Minecraft thorns.